Intermediate training draft (#644)
nextflow-io/training: docs/side_quests/working_with_files.md

nextflow run main.nf
```

```console title="Results Directory"
 N E X T F L O W   ~  version 24.10.5

Launching `./main.nf` [insane_swartz] DSL2 - revision: fff18abe6d

executor >  local (8)
[e3/449081] process > ANALYZE_READS (sampleC) [100%] 8 of 8 ✔
```

Check the results directory now:

```console title="Results Directory"
> tree results
results/
├── normal
│   ├── sampleA
│   │   ├── 1
│   │   │   └── sampleA_stats.txt
│   │   └── 2
│   │       └── sampleA_stats.txt
│   ├── sampleB
│   │   └── 1
│   │       └── sampleB_stats.txt
│   └── sampleC
│       └── 1
│           └── sampleC_stats.txt
└── tumor
    ├── sampleA
    │   ├── 1
    │   │   └── sampleA_stats.txt
    │   └── 2
    │       └── sampleA_stats.txt
    ├── sampleB
    │   └── 1
    │       └── sampleB_stats.txt
    └── sampleC
        └── 1
            └── sampleC_stats.txt
```

See how using sample metadata as values gives us powerful flexibility in our pipeline. By propagating metadata alongside our data in tuples, we can:

1. Create organized output directories based on sample attributes
2. Make decisions in processes based on sample properties
3. Split, join, and recombine data based on metadata values

> This pattern of keeping metadata explicit and attached to the data (rather than encoded in filenames) is a core best practice in Nextflow that enables building robust, maintainable bioinformatics workflows.

Reviewer: I do feel like some of this should be replaced with a link to the metadata material.

### Takeaway

> - Processes can take files as input and produce new files as output

Reviewer: ```suggestion
```
Don't think that's a main teaching point here

> - The `publishDir` directive organizes outputs based on metadata values

Reviewer: ```suggestion
- The `publishDir` directive can organize outputs based on metadata values
```

- Metadata in tuples enables structured organization of results
- This approach creates maintainable workflows with clear data provenance

---

> ## 6. Using remote files

Reviewer: As a said further up, I don't think this should be a 'by the way' at the end, but rather something we show up-front.

- HTTP(S)/FTP (http://, https://, ftp://)
- Amazon S3 (s3://)
- Azure Blob Storage (az://)
- Google Cloud Storage (gs://)

To use these, simply replace the string and Nextflow will handle authentication and staging the files to the right place, downloading or uploading and all other file operations you would expect. We call this string the Uniform Resource Identifier (URI).

The key strength of this is we can switch between environments without changing any pipeline logic. For example, you can develop with a small, local test set before moving to a remote set of data by changing the URI.

### 6.1 Using a file from the internet

!!! warning
Accessing remote data requires an internet connection!

In your workflow, replace the string path with an HTTPS one to download this file from the internet. We are going to swap the relative path of the FASTQ files with the remote one stored on the internet. This is the same data as we have been previously using.

=== "After"

    ```groovy title="main.nf" linenums="27"

>     ch_fastq = Channel.fromFilePairs('https://github.com/nextflow-io/training/blob/bb187e3bfdf4eec2c53b3b08d2b60fdd7003b763/side-quests/working_with_files/data/sampleA_rep1_normal_R1_001.fastq.gz')

Reviewer: This would need to be a glob path, no?

Reviewer: No, actually, we'd need to not use `fromFilePairs`, right?

Reviewer: Yeah, but it only produces one file:

```
[sampleA_rep1_normal_R1_001, [/nextflow-io/training/blob/bb187e3bfdf4eec2c53b3b08d2b60fdd7003b763/side-quests/working_with_files/data/sampleA_rep1_normal_R1_001.fastq.gz]]
```

Somewhat defeats the point of `fromFilePairs`

Reviewer: Yep, exactly my thinking when I suggested putting remove files earlier rather than as a bolt-on.

        [
          [
            id: sample,
            replicate: replicate,
            type: type,
            readNum: readNum,
          ],
          myFile
        ]
    }
    ```

=== "Before"

    ```groovy title="file_operations.nf" linenums="4" hl_lines="3"
    ch_fastq.map { myFile ->
        def (sample, replicate, type, readNum) = myFile.simpleName.tokenize('_')
        [ sample, replicate, type, readNum.readNum, myFile ]
    }
    ```

```bash
nextflow run file_operations.nf
```

```console title="Map Output"
 N E X T F L O W   ~  version 24.10.5

Launching `file_operations.nf` [infallible_swartz] DSL2 - revision: 7f4e68c0cb

[[id:patientA, replicate:rep1, type:normal, readNum:R2], /workspaces/training/side-quests/working_with_files/data/patientA_rep1_normal_R2_001.fastq.gz]
[[id:patientA, replicate:rep1, type:normal, readNum:R1], /workspaces/training/side-quests/working_with_files/data/patientA_rep1_normal_R1_001.fastq.gz]
```

We have converted our flat list into a map, and now we can refer to each bit of sample data by name instead of by index. This makes our code easier to read and more maintainable.

> ### 4.4. Simplifying the metadata
> 
> The metadata includes a lot of redundant information such as "rep". Let's remove this so replicate is just a number.
> 
> We can simplify the metadata by using the `.replace()` method on the replicate string to remove the "rep" prefix, and similarly remove the "R" prefix from the readNum value.
> 
> === "After"
> 
>     ```groovy title="file_operations.nf" linenums="4" hl_lines="6 8"
>     ch_fastq.map { myFile ->
>         def (sample, replicate, type, readNum) = myFile.simpleName.tokenize('_')
>         [
>           [
>             id: sample,
>             replicate: replicate.replace('rep', ''),
>             type: type,
>             readNum: readNum.replace('R', ''),
>           ],
>           myFile
>         ]
>     }
>     ```
> 
> === "Before"
> 
>     ```groovy title="file_operations.nf" linenums="4"
>     ch_fastq.map { myFile ->
>         def (sample, replicate, type, readNum) = myFile.simpleName.tokenize('_')
>         [
>           [
>             id: sample,
>             replicate: replicate,
>             type: type,
>             readNum: readNum,
>           ],
>           myFile
>         ]
>     }
>     .view()
>     ```
> 
> ```bash
> nextflow run file_operations.nf
> ```
> 
> ```console title="Simplified Metadata Output"
>  N E X T F L O W   ~  version 24.10.4
> 
> Launching `file_operations.nf` [reverent_volta] DSL2 - revision: b3aac71fea
> 
> [[sample:sampleA, replicate:1, type:normal, readNum:2], /workspaces/training/side-quests/working_with_files/data/sampleA_rep1_normal_R2_001.fastq.gz]
> [[sample:sampleA, replicate:1, type:normal, readNum:1], /workspaces/training/side-quests/working_with_files/data/sampleA_rep1_normal_R1_001.fastq.gz]
> ```
> 
> Nice! we have removed the "rep" from the replicate string.

Reviewer: ```suggestion
```
